code/krepel/memory/allocation.d

```diff
@@ -0,0 +1,691 @@
+  }
+
+  /// See_Also: krepel.system.Deallocate
+  bool Deallocate(MemoryRegion MemoryToDeallocate)
+  {
+    return SystemMemoryDeallocation(MemoryToDeallocate);
+  }
+}
+
+debug = HeapMemory;
+
+/// Can allocate arbitrary sizes of memory blocks and deallocate them in any
+/// order.
+///
+/// Uses an implicit free list and a first-fit for allocations.
+///
+/// Usually a block has a boundary tag at the end to indicate the size of that
+/// block. This is done to enable bidirectional traversal from any given
+/// block. This implementation, however, does not require such a boundary tag.
+/// When allocating memory, a linear search is performed, starting from the
+/// first block. Along the way, all adjacent free blocks are merged, until a
+/// suitable block is found that satisfies the block size requirements of that
+/// allocation request. Thus, we lazily merge adjacent free blocks on demand
+/// only.
+///
+/// Layout of a memory block:
+/// [********][??????????*][.??????????????]
+///  \ Tag  /  \ Padding /  \ User Memory /
+///
+/// * => Reserved memory, implementation overhead.
+/// ? => Unknown number of bytes
+/// . => Byte used by the user
+///
+/// The `Tag` size is exactly size_t.sizeof bytes. It contains the size of the
+/// block and a flag whether this block is allocated or not.
+///
+/// Padding is >= 1 byte. The byte to the far right encodes a ubyte value that
+/// states the actual padding size. This value is used to free memory in an
+/// efficient way while guaranteeing alignment requirements.
+struct HeapMemory
+{
+  mixin MemoryMixinTemplate!(SupportsDeallocation);
+  mixin MemoryContainsCheckMixin!(Memory);
+
+  MemoryRegion Memory;
+  size_t DefaultAlignment = GlobalDefaultAlignment;
+
+  debug(HeapMemory) @property bool IsInitialized() const { return cast(bool)FirstBlock; }
+
+  /// The number of bytes needed by a block header.
+  enum BlockOverhead = size_t.sizeof;
+
+  /// The number of bytes required to make use of a block in the case the user
+  /// requests 1 byte of memory with an alignment of 1.
+  enum MinimumBlockSize = BlockOverhead + 1;
+
+
+  this(MemoryRegion AvailableMemory)
+  {
+    Initialize(AvailableMemory);
+  }
+
+  void Initialize(MemoryRegion AvailableMemory)
+  {
+    debug(HeapMemory)
+    {
+      assert(!IsInitialized, "Heap memory must not be initialized.");
+      assert(DefaultAlignment.IsPowerOfTwo, "DefaultAlignment is expected to be a power of two.");
+    }
+
+    Memory = AvailableMemory;
+
+    auto BlockSize = Memory.length;
+    if(BlockSize.IsOdd) BlockSize--;
+    FirstBlock = cast(BlockData*)Memory.ptr;
+    FirstBlock.Size = BlockSize;
+    FirstBlock.IsAllocated = false;
+  }
+
+  auto Allocate(size_t RequestedBytes, size_t Alignment = 0)
+  {
+    debug(HeapMemory)
+    {
+      assert(IsInitialized, "Heap memory is not initialized.");
+      assert(Alignment == 0 || Alignment < ubyte.max,
+             "Alignment value is supposed to fit into 1 byte.");
```

Comment:
Especially on consoles the alignment sizes can be up to 2MB! (Not necessary for now)